integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala

```diff
@@ -330,6 +207,264 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
+      // in case of average child table will have two columns which will be stored in sequence
+      // so for average expression we need to get two columns for mapping
+      var counter = 0
+      // sorting the columns based on schema ordinal so search will give proper result
+      val sortedColumnList = aggDataMapSchema.getChildSchema.getListOfColumns.asScala
+        .sortBy(_.getSchemaOrdinal)
+      val logicalPlanToColumnMapping = dataMapAggExp.map { aggExp =>
+        // for each aggregate expression get logical plan
+        val expLogicalPlan = getLogicalPlanFromAggExp(aggExp,
+          carbonTable.getTableName,
+          carbonTable.getDatabaseName, logicalRelation)
+        // check if aggregate expression is of type avg
+        // get the columns
+        var columnSchema = aggDataMapSchema
+          .getAggColumnBasedOnIndex(counter, sortedColumnList.asJava)
+        // increment the counter so when for next expression above code will be
+        // executed it will search from that schema ordinal
+        counter = columnSchema.getSchemaOrdinal + 1
+        (expLogicalPlan, columnSchema)
+      }.toMap
+      // store the mapping in data map
+      aggDataMapSchema.setAggregateExpressionToColumnMapping(logicalPlanToColumnMapping.asJava)
+      // return the mapping
+      logicalPlanToColumnMapping
+    } else {
+      // if already present in data map then return the same
+      aggDataMapSchema.getAggregateExpressionToColumnMapping
+        .asInstanceOf[java.util.Map[LogicalPlan, ColumnSchema]].asScala.toMap
+    }
+  }
+
+
+  /**
+   * Below method will be used to get the logical plan from aggregate expression
+   * @param aggExp
+   *               aggregate expression
+   * @param tableName
+   *                  parent table name
+   * @param databaseName
+   *                     database name
+   * @param logicalRelation
+   *                        logical relation
+   * @return logical plan
+   */
+  def getLogicalPlanFromAggExp(aggExp: AggregateExpression,
+      tableName: String,
+      databaseName: String,
+      logicalRelation: LogicalRelation): LogicalPlan = {
+    // adding the preAGG UDF, so pre aggregate data loading rule and query rule will not
+    // be applied
+    val query = new CarbonSpark2SqlParser()
+      .addPreAggFunction(s"Select ${ aggExp.sql } from $databaseName.$tableName")
+    // updating the logical relation of logical plan to so when two logical plan
+    // will be compared it will not consider relation
+    updateLogicalRelation(sparkSession.sql(query).logicalPlan, logicalRelation)
+  }
+
+  /**
+   * Below method will be used to get aggregate expression
+   * @param logicalPlan
+   *               logical plan
+   * @return list of aggregate expression
+   */
+  def getAggregateExpFromChildDataMap(logicalPlan: LogicalPlan): Seq[AggregateExpression] = {
+    val list = scala.collection.mutable.HashSet.empty[AggregateExpression]
```

Comment:
Should use List or LinkedHasSet as the order of insertion is required for mapping